From 113142939172bda9759d4d2e5a053911821d2faf Mon Sep 17 00:00:00 2001
From: "t.feng" <fengtao40@huawei.com>
Date: Thu, 10 Nov 2022 15:01:15 +0800
Subject: loader/multiboot_elfxx: Fix memory leak

The commit eb33e61b3 (multiboot: fix memory leak) did not fix all
issues. Fix all of them right now.

Fixes: eb33e61b3 (multiboot: fix memory leak)

Reference:https://git.savannah.gnu.org/cgit/grub.git/commit/?id=113142939172bda9759d4d2e5a053911821d2faf
Conflict:NA

Signed-off-by: t.feng <fengtao40@huawei.com>
Reviewed-by: Daniel Kiper <daniel.kiper@oracle.com>
---
 grub-core/loader/multiboot_elfxx.c | 23 ++++++++++++++---------
 1 file changed, 14 insertions(+), 9 deletions(-)

diff --git a/grub-core/loader/multiboot_elfxx.c b/grub-core/loader/multiboot_elfxx.c
index 87f6e31..57330a0 100644
--- a/grub-core/loader/multiboot_elfxx.c
+++ b/grub-core/loader/multiboot_elfxx.c
@@ -220,10 +220,7 @@ CONCAT(grub_multiboot_load_elf, XX) (mbi_load_data_t *mld)
 	return grub_errno;
       
       if (grub_file_seek (mld->file, ehdr->e_shoff) == (grub_off_t) -1)
-	{
-	  grub_free (shdr);
-	  return grub_errno;
-	}
+	goto fail;
 
       if (grub_file_read (mld->file, shdr, (grub_uint32_t) ehdr->e_shnum * ehdr->e_shentsize)
               != (grub_ssize_t) ehdr->e_shnum * ehdr->e_shentsize)
@@ -231,7 +228,7 @@ CONCAT(grub_multiboot_load_elf, XX) (mbi_load_data_t *mld)
 	  if (!grub_errno)
 	    grub_error (GRUB_ERR_FILE_READ_ERROR, N_("premature end of file %s"),
 			mld->filename);
-	  return grub_errno;
+	  goto fail;
 	}
       
       for (shdrptr = shdr, i = 0; i < ehdr->e_shnum;
@@ -242,7 +239,10 @@ CONCAT(grub_multiboot_load_elf, XX) (mbi_load_data_t *mld)
 	  grub_addr_t target;
 
 	  if (mld->mbi_ver >= 2 && (sh->sh_type == SHT_REL || sh->sh_type == SHT_RELA))
-	    return grub_error (GRUB_ERR_NOT_IMPLEMENTED_YET, "ELF files with relocs are not supported yet");
+	    {
+	      grub_error (GRUB_ERR_NOT_IMPLEMENTED_YET, "ELF files with relocs are not supported yet");
+	      goto fail;
+	    }
 
 	  /* This section is a loaded section,
 	     so we don't care.  */
@@ -261,13 +261,14 @@ CONCAT(grub_multiboot_load_elf, XX) (mbi_load_data_t *mld)
 	  if (err)
 	    {
 	      grub_dprintf ("multiboot_loader", "Error loading shdr %d\n", i);
-	      return err;
+	      grub_errno = err;
+	      goto fail;
 	    }
 	  src = get_virtual_current_address (ch);
 	  target = get_physical_target_address (ch);
 
 	  if (grub_file_seek (mld->file, sh->sh_offset) == (grub_off_t) -1)
-	    return grub_errno;
+	    goto fail;
 
           if (grub_file_read (mld->file, src, sh->sh_size)
               != (grub_ssize_t) sh->sh_size)
@@ -275,12 +276,16 @@ CONCAT(grub_multiboot_load_elf, XX) (mbi_load_data_t *mld)
 	      if (!grub_errno)
 		grub_error (GRUB_ERR_FILE_READ_ERROR, N_("premature end of file %s"),
 			    mld->filename);
-	      return grub_errno;
+	      goto fail;
 	    }
 	  sh->sh_addr = target;
 	}
       GRUB_MULTIBOOT (add_elfsyms) (ehdr->e_shnum, ehdr->e_shentsize,
 				    ehdr->e_shstrndx, shdr);
+      return GRUB_ERR_NONE;
+
+fail:
+      grub_free (shdr);
     }
 
 #undef phdr
-- 
2.19.1

